### ISSUE TITLE
Explicitly spec out "native origin" behind each XRSpace

### ISSUE THREAD

**Issue by Manishearth** [immersive-web/webxr#597]
@NellWaliczek brought this up in the meeting

Basically, we should explicitly specify an internal "natural origin" or "pose" ("transform"?) related to each kind of XR Space. This transform is relative to some implementation-defined internal global origin consistent across spaces (what this origin is doesn't really matter since it's never exposed -- users can only query the relationship between two spaces).

This should clarify https://github.com/immersive-web/webxr/issues/565 and https://github.com/immersive-web/webxr/issues/581 and a bunch of other issues that are floating around.

I don't mind working on this if folks think this is valuable to try (perhaps worth waiting for https://github.com/immersive-web/webxr/pull/587 and https://github.com/immersive-web/webxr/pull/589 first?)

**Comment by Manishearth** [comment]
To give an idea of what I'm suggesting, the spec text for this could be a more precise version of the following (filled out to be more precise, and filed under the appropriate XRSpace types)

 - The natural origin of viewerSpace and (zero-offset) identity spaces is the position and orientation of the user's head.
 - The natural origin of eye-level is the position and orientation of the user's head when the session starts
 - The natural origin of floor-level is more or less the position and orientation of the user's feet at session creation time (which can also be approximated by computing `standing_to_sitting * eye_level_origin`
 - The natural origin of an offset reference space is `unoffset_origin * offset` (I'm not sure if we've 100% settled on what the multiplication order is here, cc @klausw )
 - The natural origin of an input space is the position and orientation of that controller
 - The natural origin of a bounded reference space is at the floor-level at the center of the room, with an orientation depending on the underlying platform preferences
 - The natural origin of an unbounded reference space is somewhere near the user's head at the time of session creation, but is allowed to drift over time.

We can also add explicit language mentioning how things are expected to work in case of the limits of 3DOF devices (i.e. the `eye_level` origin will follow the headset around if the user walks). Alternatively we can at the top of the spec mention what the expectations are for 3DOF devices ("assume that no walking has occurred")

We can now specify `getPose(dest, source)` as `source_natural_origin^-1 * dest_natural_origin`, and `getViewerPose(space)` as `getPose(dest=viewerSpace, source=space) = space_natural_origin^-1 * viewerSpace_natural_origin`

**Comment by Manishearth** [comment]
Specifying unbounded spaces is tricky in this model, because what we really want is items that travel with the user to continue to report high-fidelity `getPose()` values, but other reference spaces should be free to drift. One way to look at this is that the "implementation-defined internal global origin" that all spaces are relative to is what's allowed to drift, and it is a user error to try and use `stationary` or `bounded` reference spaces in such a context (so `getPose(stationary, unbounded)` will likely return some fixed value that is not required to make sense in this model). A cleaner way of specifying this is that the "implemetation-defined internal global origin" can be something that is relative to the room or viewer, based on the type of experience at hand, and use a bunch of non-normative notes to clarify what's going on. Ultimately the "implemetation-defined internal global origin" concept is mostly so that we can use matrix math (after all, implementations will ultimately have something like this!) cleanly.

**Comment by cwilso** [comment]
This should be done after #587 (builds on top of that work).

**Comment by Manishearth** [comment]
Yep, that's the plan. See https://github.com/immersive-web/editor-collab/pull/38#issuecomment-487754222

**Comment by thetuvix** [comment]
Thanks for driving these space definitions!  The model of just mapping each space's origin to some physically-based definition is definitely the way to go - any alternative that defines some custom method to calculate poses for any particular pair of spaces will quickly lead to madness :smile:

@Manishearth:
> so getPose(stationary, unbounded) will likely return some fixed value that is not required to make sense in this model

Can you elaborate on why `unbounded` would need to be specified any differently?  Specifically, what would cause `getPose(stationary, unbounded)` to be fixed, given that the purpose of an `unbounded` space is to drift over time relative to anchored spaces like `stationary`?

Per https://github.com/immersive-web/webxr/pull/609/files#diff-ec9cfa5f3f35ec1f84feb2e59686c34dR713:
> An `XRSpace` represents a virtual coordinate system with an origin that corresponds to a physical location. Spatial data that is requested from the API or given to the API is always expressed in relation to a specific `XRSpace` at the time of a specific `XRFrame`. Numeric values such as pose positions are coordinates in that space relative to its origin. The interface is intentionally opaque.

If you asked `getPose(stationary, unbounded)`, the transform you get would be the transform that gets you from "stationary space" to "unbounded space".  This should be precisely the difference in transform between asking `getPose(anchor, stationary)` and `getPose(anchor, unbounded)`, or equivalently the difference in transform between asking `getViewerPose(stationary)`, `getViewerPose(unbounded)`.

The result of the exercise you're doing here should be that 100% of `getPose` requests are defined simply as the transform between the two physical locations in the real world represented by each `XRSpace` - full stop.

For what it's worth, neither the Windows MR API nor the OpenXR API have any developer-facing notion of "implementation-defined internal global origin".  You simply ask how two spaces you care about relate to one another, be they viewer spaces, anchor spaces, reference spaces, etc. (just like you do as a WebXR developer!)  UAs may ultimately pick some global substrate space to optimize how they communicate poses across processes, but we can define correctness in the WebXR spec itself purely in physical terms.

Note that the "unbounded" space as WebXR defines it is actually a great substrate space for that purpose, since, per #559, it will always be locatable after the device initially establishes positional tracking.  Microsoft will be proposing an equivalent of the "unbounded" space as an OpenXR extension for platforms that support unbounded tracking, which should make a WebXR implementation on top of OpenXR even easier:
* If the OpenXR `UNBOUNDED_EXT` reference space can be created, the UA can bind that directly to the WebXR `unbounded` reference space.  A UA could then also use this underlying OpenXR `UNBOUNDED_EXT` space as its global substrate space if it so chooses.
* If the OpenXR `UNBOUNDED_EXT` reference space cannot be created, the UA can disallow creation of the WebXR `unbounded` reference space.  A UA could then also use the underlying OpenXR `LOCAL` space as its global substrate space, which will otherwise bind directly to the WebXR `stationary`/`eye-level` reference space.  The additional positional tracking capability of `UNBOUNDED_EXT` vs. `LOCAL` is to give positions when the device regains tracking in a second disjoint room - however, that won't occur on devices that do not support unbounded tracking, and so `LOCAL` would be sufficient.

@Manishearth:
> We can also add explicit language mentioning how things are expected to work in case of the limits of 3DOF devices (i.e. the eye_level origin will follow the headset around if the user walks). Alternatively we can at the top of the spec mention what the expectations are for 3DOF devices ("assume that no walking has occurred")

We could simply define that the origin of viewer space for a device without positional tracking coincides with the natural origin of the stationary eye-level reference space. A UA can then return a position that is slightly off from that neutral position due to neck modeling, etc - the rest of the "follow the headset around" rules should just fall out from the general tracking loss rules that we spec from #559.

**Comment by Manishearth** [comment]
Note: the comments in this issue have since been superseded by a PR at https://github.com/immersive-web/editor-collab/pull/38, so some of the stuff here is old

> Can you elaborate on why unbounded would need to be specified any differently? Specifically, what would cause getPose(stationary, unbounded) to be fixed, given that the purpose of an unbounded space is to drift over time relative to anchored spaces like stationary?

I think I misstated that, I should not have said "fixed". I was trying to comment on the exact way to spec the "drift over time" thing. Because while we'd like to allow the origin to drift over time wrt stationary origins, we do not want it to drift over time wrt the viewer and any inputs that travel with the viewer (aside from changes caused by the motion of the viewer or controller). So if the viewer walks far away from  the starting point and then stops moving, `getViewerPose(unbounded)` shouldn't stay the same after they stop moving. If the user is carrying an input, `getPose(input, viewer)` should still be accurate. However `getPose(unbounded, stationary)` is allowed to drift.

What's actually happening is that viewer, viewer-carried input, and unbounded spaces are all allowed to have their reported origins drift over time in the _same way_ from their real-world origins.

I don't think we _have_ to get into this in the spec, it was just a side issue that was made more apparent by trying to talk concretely in terms of origins.


> We could simply define that the origin of viewer space for a device without positional tracking coincides with the natural origin of the stationary eye-level reference space

That works.
